autogpt_platform/backend/backend/util/service.py

```diff
@@ -273,10 +273,19 @@ def __start_fastapi(self):
         )
         self.shared_event_loop.run_until_complete(server.serve())
 
+    def cleanup(self):
+        """Override to set shutdown flag during cleanup phase."""
+        logger.info(f"[{self.service_name}] 🛑 Entering graceful shutdown")
+        self._shutting_down = True
+        super().cleanup()
+
     async def health_check(self) -> str:
         """
         A method to check the health of the process.
+        Fails during shutdown to prevent k8s from routing new requests.
         """
+        if self._shutting_down:
+            raise UnhealthyServiceError("Service is shutting down")
```

Comment:
Does service shutting down always means it's not healthy? or is this only for database manager?

Reply:
I think this applies to every `AppService` with RPC endpoints. After they get a shutdown signal, they do cleanup things that put the service in a broken state, meaning they likely can't process new requests.

Reply:
I've checked the logs and the time between termination signal and the endpoint error is generally < 100 ms, which leads me to believe the errors is actually due to the process being terminated, not due to its cleanup. We should probably delay shutdown by e.g. 30 seconds so it can finish processing its requests and giving the load balancer time to deregister the pod as well so it doesn't receive new requests.

Reply:
https://github.com/Significant-Gravitas/AutoGPT_cloud_infrastructure/pull/251